feature: add runtime support for private_key_jwt client authentication

strehle opened the pull request:
More details in https://github.com/cloudfoundry/uaa/pull/2449, in #2433 as this PR include #2433. -> because to have smaller review packages

Enable the validation of client_assertion as replacement for client_secret Add private_key_jwt as client_auth_method into tokens.

strehle committed:
feature: add runtime support for private_key_jwt client authentication

More details in https://github.com/cloudfoundry/uaa/pull/2449, in #2433 as this PR include #2433. -> because to have smaller review packages

Enable the validation of client_assertion as replacement for client_secret
Add private_key_jwt as client_auth_method into tokens.

strehle commented:
last PR for this story, after that private_key_jwt works.

Will provide with next PR more integration tests and examples

adrianhoelzl-sap commented on server/src/main/java/org/cloudfoundry/identity/uaa/client/UaaClient.java:
```diff
@@ -46,6 +52,12 @@ private Map<String, Object> getAdditionalInformation() {
+  public ClientJwtConfiguration getClientJwtConfiguration() {
+    UaaClientDetails uaaClientDetails = new UaaClientDetails();
+    uaaClientDetails.setClientJwtConfig(clientJwtConfig);
+    return Optional.ofNullable(ClientJwtConfiguration.readValue(uaaClientDetails)).orElse(new ClientJwtConfiguration());
```
Consider replacing this with `JsonUtils.readValue(clientDetails.getClientJwtConfig(), ClientJwtConfiguration.class)` or add a separate `readValue` method to `ClientJwtConfiguration` that takes the String as a parameter

strehle replied:
done

adrianhoelzl-sap commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthentication.java:
```diff
@@ -65,4 +104,61 @@ private static HashMap<String, String> getJwtClientConfigurationElements(Object
+        }
+        JWT clientToken = JWTParser.parse(UaaStringUtils.getSafeParameterValue(requestParameters.get(CLIENT_ASSERTION)));
+        return clientId.equals(validateClientJWToken(clientToken, JWKSet.parse(
+            oidcMetadataFetcher.fetchWebKeySet(clientJwtConfiguration).getKeySetMap()), clientId, keyInfoService.getTokenEndpointUrl()).getSubject());
```
If the first constructor is used, `oidcMetadataFetcher` is null at this point

adrianhoelzl-sap commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthentication.java:
```diff
@@ -65,4 +104,61 @@ private static HashMap<String, String> getJwtClientConfigurationElements(Object
+        if (!clientId.equals(getClientId(clientAssertion))) {
+          throw new BadCredentialsException("Wrong client_assertion");
+        }
+        JWT clientToken = JWTParser.parse(UaaStringUtils.getSafeParameterValue(requestParameters.get(CLIENT_ASSERTION)));
```
```suggestion
        JWT clientToken = JWTParser.parse(clientAssertion);
```

adrianhoelzl-sap replied:
Can we reuse the `clientAssertion` variable here?

strehle replied:
done

adrianhoelzl-sap commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthentication.java:
```diff
@@ -65,4 +104,61 @@ private static HashMap<String, String> getJwtClientConfigurationElements(Object
+    ConfigurableJWTProcessor<SecurityContext> jwtProcessor = new DefaultJWTProcessor<>();
+    jwtProcessor.setJWSKeySelector(keySelector);
+
+    Set<String> intRequiredClaims = new HashSet<>(
+        Arrays.asList(ClaimConstants.ISS, ClaimConstants.SUB, ClaimConstants.AUD, ClaimConstants.EXPIRY_IN_SECONDS, ClaimConstants.JTI));
```
Can we make this a constant?

strehle replied:
done

adrianhoelzl-sap commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthenticationTest.java:
```diff
@@ -189,6 +200,92 @@ void testGetClientAssertionCustomSingingKeyButNoCertificate() throws ParseExcept
+    // Then
+    Exception exception = assertThrows(BadCredentialsException.class, () ->
+        jwtClientAuthentication.validateClientJwt(getMockedRequestParameter(null,  jwtClientAuthentication.getClientAssertion(config)),
+            getMockedClientJwtConfiguration(null), "test"));
```
Please comment here that the parameter `"test"` should cause the exception

adrianhoelzl-sap commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthenticationTest.java:
```diff
@@ -189,6 +200,92 @@ void testGetClientAssertionCustomSingingKeyButNoCertificate() throws ParseExcept
+
+  @Test
+  void testInvalidateClientAssertion() throws Exception {
+    // Given
+    jwtClientAuthentication = new JwtClientAuthentication(keyInfoService, oidcMetadataFetcher);
+    ClientJwtConfiguration clientJwtConfiguration = getMockedClientJwtConfiguration("extId");
+    when(oidcMetadataFetcher.fetchWebKeySet(clientJwtConfiguration)).thenReturn(clientJwtConfiguration.getJwkSet());
+    // When
+    String clientAssertion = (String) jwtClientAuthentication.getClientAssertion(config);
+    // Then
+    Exception exception = assertThrows(BadCredentialsException.class, () ->
+        jwtClientAuthentication.validateClientJwt(getMockedRequestParameter(null, clientAssertion), clientJwtConfiguration, "identity"));
+    assertEquals("Untrusted client_assertion", exception.getMessage());
```
Please comment here how the expected exception is caused

strehle replied:
done

adrianhoelzl-sap commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthenticationTest.java:
```diff
@@ -189,6 +200,92 @@ void testGetClientAssertionCustomSingingKeyButNoCertificate() throws ParseExcept
+
+  @Test
+  void testSignatureInvalidateClientAssertion() throws Exception {
+    // Given
+    jwtClientAuthentication = new JwtClientAuthentication(keyInfoService, oidcMetadataFetcher);
+    ClientJwtConfiguration clientJwtConfiguration = getMockedClientJwtConfiguration(null);
+    when(oidcMetadataFetcher.fetchWebKeySet(clientJwtConfiguration)).thenReturn(clientJwtConfiguration.getJwkSet());
+    // When
+    String clientAssertion = (String) jwtClientAuthentication.getClientAssertion(config);
+    // Then
+    Exception exception = assertThrows(BadCredentialsException.class, () ->
+        jwtClientAuthentication.validateClientJwt(getMockedRequestParameter(null, clientAssertion), clientJwtConfiguration, "identity"));
+    assertEquals("Unauthorized client_assertion", exception.getMessage());
```
Same here, please comment how the exception is caused

strehle replied:
done

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into feature/issue/2235/jwtRuntime

strehle committed:
review

strehle committed:
fix smells from sonar

https://sonarcloud.io/summary/new_code?id=cloudfoundry-identity-parent&pullRequest=2507

strehle commented:
Sonar Results

https://sonarcloud.io/summary/new_code?id=cloudfoundry-identity-parent&pullRequest=2507

strehle commented:
@Tallicia FYI, this is the final PR for private_key_jwt runtime, so please ask someone in your team to review it from vmware side.

There are also PRs for integration testing and documentation, but this only the server side authentication
